ISO15693 Support added

sgtJohnny opened the pull request:
I've added basic ISO15693 support to read a tag and output to Serial. Tested on Arduino UNO and CLRC663 Evaluation board.

Please note that ISO15693 only works with CLRC663, which uses same Software as MFRC630, but can read ISO15693, ISO14443B, and many more.

sgtJohnny committed:
ISO15693 Support added

Basic ISO15693 Tag reading and dump to Serial. Also Example File for Arduino added

iwanders commented on mfrc630.c:
```diff
@@ -931,3 +950,149 @@ void mfrc630_MF_example_dump() {
+
+	//transfer UID to variable
+	mfrc630_read_fifo(uid,fifo_len);
+	return 0;										//return state - valid
```
Should we return the length of the uuid here? so `fifo_len` and return negative values as errors? That's more or less in line with the other `select` function from https://github.com/iwanders/MFRC630/blob/master/mfrc630.h#L578

iwanders commented on mfrc630.h:
```diff
@@ -692,6 +692,29 @@ uint8_t mfrc630_MF_write_block(uint8_t block_address, const uint8_t* source);
+/*! \Reads ISO15693 tag
+
+  Try to Read ISO15693 tag
+*/
```
Can we add some more documentation here? What does the returned value mean?

iwanders replied:
This is much better, please add a space between `10` and `bytes`, other than that this is clear now. :+1:

iwanders commented on mfrc630.h:
```diff
@@ -692,6 +692,29 @@ uint8_t mfrc630_MF_write_block(uint8_t block_address, const uint8_t* source);
+*/
+/*! \Initializes the reader for iso15693
+
+  Set all registers to the desired value for ISO15693 reading
```
Can we add some details about possible `protocol` and `buf` values? Like;
```
mfrc630_ISO15693_init(MFRC630_PROTO_ISO15693_1_OF_4_SSC, MFRC630_RECOM_15693_ID1_SSC26)
```
Is used in the example?

iwanders commented on examples/arduino_example/arduino_example_15693.ino:
```diff
@@ -0,0 +1,130 @@
+  SPI.begin();
+
+  // Set the registers of the MFRC630 into the default.
+ mfrc630_AN1102_recommended_registers(MFRC630_PROTO_ISO15693_1_OF_4_SSC);
```
Nitpick; Indent doesn't match scope.

iwanders commented on examples/arduino_example/arduino_example_15693.ino:
```diff
@@ -0,0 +1,130 @@
+			Serial.println("Failure, No Tag found or Reader Problem!");
+			break;
+	}
+}
```
Nitpick, this function has tabs, not spaces.

iwanders reviewed (commented):
Hey @sgtJohnny, sorry for the delayed feedback, I didn't get an email from github for some reason. Thanks so much for getting back to this and contributing it as a PR. I'm sure lots of people will be grateful.

I can't do any real tests on this as I don't have the hardware anymore, so I'm limited to commenting based on the code. I trust the code works :+1:

sgtJohnny committed:
Various corrections

-Return uid lenght as status and zero as error
- adding comments
- changed arduino example

sgtJohnny commented:
Added changes you advised, please review

iwanders commented on mfrc630.c:
```diff
@@ -931,3 +950,149 @@ void mfrc630_MF_example_dump() {
+	    if(millis()>(timeout+50)){
+	    	break;
+	    }
+	  }
```
Any chance you could change this timeout handling to use the internal timer? like how's done in `mfrc630_iso14443a_WUPA_REQA`? this file is independent of Arduino at the moment, and I'd like to keep it that way.
This function has both spaces and tables on a single line, please make it uniform and match the already existing indentations.

iwanders commented on mfrc630.c:
```diff
@@ -23,6 +23,7 @@
 */
 
 #include "mfrc630.h"
+#include "Arduino.h"
```
I don't want this dependency here, it rules out using this driver outside of the arduino ecosystem.

iwanders commented on mfrc630.h:
```diff
@@ -692,6 +692,38 @@ uint8_t mfrc630_MF_write_block(uint8_t block_address, const uint8_t* source);
+  Set all registers to the desired value for ISO15693 reading
+
+   \param [in] the specified protocol number for ISO15693, there are differnt values for 1/4 SSC or DSC or 1/256 SSC.
+   The right protocol for the desired card standart must be used
```
Spelling mistakes in `difernt` and `standart`.

sgtJohnny commented:
I will try to fix this as soon as possible and try with timer

ic-twist commented:
Posting here too for visibility...

@sgtJohnny Did you or anyone else use NXP's CLEV6630B evaluation board or CLEV6630ARD board for testing and verification of your software?

**My hardware setup**
NXP CLEV6630B eval board connected to Arduino Mega Rev3 through SPI interface. 

**Software setup**
Using @sgtJohnny 's `iso15693` branch, modified SPI clock rate from 10MHz to 5MHz as shown below. As 10MHz was too fast for my Arduino to drive good SPI signals.
`SPI.beginTransaction(SPISettings(5000000, MSBFIRST, SPI_MODE0));`

**Problem**
Not able to read ISO15693 NFC tags.

**What I've Tried**
1. I used a logic analyzer and saw correct bytes (appears to match the firmware) are sending over SPI interface from Arduino Mega to the CLRC663 chip.
2. I also tried adding the command to turn on RF field in the setup. Please see my setup code below:
```
void setup()
{
  // Start serial communication.
  Serial.begin(9600);

  // Set the chip select pin to output.
  pinMode(CHIP_SELECT, OUTPUT);
  digitalWrite(CHIP_SELECT, HIGH);

  // Start the SPI bus.
  SPI.begin();
  delay(1000);
  const uint8_t ic_version = mfrc630_read_reg(MFRC630_REG_VERSION);
  Serial.println(ic_version);

  // Set the registers of the MFRC630 into the default.
  mfrc630_AN1102_recommended_registers(MFRC630_PROTO_ISO15693_1_OF_4_SSC);

  // RF on
  mfrc630_read_reg(MFRC630_REG_DRVMOD);
  mfrc630_write_reg(MFRC630_REG_DRVMOD, 0x89);
}
```
3. I compared the SPI traffic generated by the Arduino code with the SPI traffic generated by the NXP Cockpit GUI software and they don't seem to match up.

**My Speculations**
I'm wondering if the command to read ISO15693 inventory was implemented correctly.

ic-twist commented on mfrc630.c:
```diff
@@ -931,3 +950,149 @@ void mfrc630_MF_example_dump() {
   }
 }
+
+void mfrc630_ISO15693_init(uint8_t protocol, uint8_t buf){
```
@sgtJohnny I think there's a bug here - `buf` should be a `uint8_t*` type. This also means that your implementation on line 972 `mfrc630_AN1102_recommended_registers(buf);` does not do what you expect it to.

sgtJohnny commented:
@ic-twist  I remember using a blue board, but not sure if it was CLEV6630B. I've been using the code in a few own designs with the CRLC663 and ISO15693 and never had problems, but I've added a loop to try all ISO15693 settings. 

I tried to dig up the board this weekend an try with arduino uno.

ic-twist commented:
I fixed the bug in `mfrc630_ISO15693_init` function that I commented on and it started working for me. Without this change, the registers are not properly loaded. I can make a PR for my changes early next week.

CLEV6630B is also blue so it might be what you used too.

Now that it's working, I noticed that the NFC read range is much smaller (~1cm) when I'm using the arduino as a host micro compared to when I use the LPC micro on the eval board itself and use NXP cockpit GUI to read tags, in which case the range is ~1 inch.
Do you know what can improve the read range?

ic-twist commented:
Ok - I needed to change some of the register values. The one that made a big difference was `MFRC630_REG_RXANA` which sets the gain.
Also in the process of debugging I realized that I needed to change one of the jumper wires to make sure the CLRC output Vdd is set to 5V. Thanks to this post https://community.nxp.com/t5/NFC/CLRC663-Programming-Guide-via-UART/m-p/664722.

@sgtJohnny The software bug I mentioned is still necessary for this library to work.
